### DISCUSSION TITLE
Clarification on the full scope allowed switch

### DISCUSSION THREAD

**Discussion by yann-soubeyrand** [keycloak/keycloak#20006] [Q&A] [answered]
Hello,

It’s not very clear in my mind what the “full scope allowed” switch in the client dedicated scopes settings does. When ON, and the user has the `manage-account` role, the application gets an access token with `account` in the audience claim. Does it mean that this application can now use this access token to manage the user account?

**Comment by jbman** [comment, marked as answer]
The switch allows to create access tokens with only a subset of the user's roles. You probably want to set it to "off" and configure all roles which are actually needed by the client. 
And yes, if the `manage-account-role `is in the token it can be used at the account API for user-self-service operations. So if your client doesn't do sth. with this you can achieve a token with less privileges if you switch "Full scope allowed" to "off" and configure needed roles.

**Reply by jbman** [reply]
Maybe also see this discussion: https://github.com/keycloak/keycloak/discussions/19869, section: "Dedicated client scope and "Full scope allowed" setting"

**Comment by jonkoops** [comment]
@xianli123 perhaps we can improve the copy for this?

**Reply by xianli123** [reply]
@jonkoops Yes, we can update this part and then ask Stian or others to review it. I am not sure if there are any new decisions for the client scopes inside the Clients.

**Reply by xianli123** [reply]
I will provide some designs for you. @jonkoops

**Comment by yann-soubeyrand** [comment]
Thanks for your answer @jbman.

I’m not an OAuth2/OpenID Connect expert, but, although this default “Full scope allowed” setting seems appropriate for an ecosystem where a single frontend application can make API calls to multiple trusted backend services on behalf of the user, it seems pretty insecure in presence of untrusted applications (like explained in the documentation). Maybe it should be clarified why this is the default setting and it should be emphasized that it’s strongly advised to change this setting (or maybe I missed some point here)?

**Reply by jbman** [reply]
I agree that it would be much better to have secure by default settings. However it is hard to change existing defaults without breaking things. 
The UI may be improved to support a more secure configuration. Maybe in future the admin console could introduce "off" by default.

My proposal is a tab named "Delegated Roles". Ar text should explains that you need to add some roles here, in order to delegate then to the client in the access token. If "full scope allowed" is switched off, it should also be stated that no roles will be put into the access token.
